finality-aleph/src/new_network/session.rs

```diff
@@ -0,0 +1,115 @@
+use crate::{
+    crypto::{AuthorityPen, AuthorityVerifier},
+    new_network::{BareNetwork, SendError, Sender as SenderTrait, SessionCommand},
+    NodeIndex, SessionId,
+};
+use aleph_bft::Recipient;
+use codec::Codec;
+use futures::channel::mpsc;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+/// Sends data within a single session.
+#[derive(Clone)]
+pub struct Sender<D: Clone + Codec + Send> {
+    session_id: SessionId,
+    messages_for_network: mpsc::UnboundedSender<(D, SessionId, Recipient)>,
+}
+
+impl<D: Clone + Codec + Send> SenderTrait<D> for Sender<D> {
+    fn send(&self, data: D, recipient: Recipient) -> Result<(), SendError> {
+        self.messages_for_network
+            .unbounded_send((data, self.session_id, recipient))
+            .map_err(|_| SendError::SendFailed)
+    }
+}
+
+/// Sends and receives data within a single session.
+pub struct Network<D: Clone + Codec + Send> {
+    sender: Sender<D>,
+    receiver: Arc<Mutex<mpsc::UnboundedReceiver<D>>>,
+}
+
+impl<D: Clone + Codec + Send> BareNetwork<D> for Network<D> {
+    type S = Sender<D>;
+    type R = mpsc::UnboundedReceiver<D>;
+    fn sender(&self) -> &Self::S {
+        &self.sender
+    }
+    fn receiver(&self) -> Arc<Mutex<Self::R>> {
+        self.receiver.clone()
+    }
+}
+
+/// Manages sessions for which the network should be active.
+pub struct Manager<D: Clone + Codec + Send> {
+    commands_for_service: mpsc::UnboundedSender<SessionCommand<D>>,
+    messages_for_service: mpsc::UnboundedSender<(D, SessionId, Recipient)>,
+}
+
+/// What went wrond during a session management operation.
+pub enum ManagerError {
+    SendCommand,
```

Comment:
If I would see this error message I would be at least confused

Reply:
Changed to `CommandSendFailed`.